### PR TITLE
Update pyarrow requirement from <8.1.0,>=8.0.0 to >=10.0.1,<10.1.0

### PR DESCRIPTION
**Pull request by 3dbrows**
This is intended to replace the dependabot PR #1304 by also updating pyproject.toml, and fixing us to one major version.

I am not too sure how to test this against the C++ [ArrowIterator](https://github.com/snowflakedb/snowflake-connector-python/tree/main/src/snowflake/connector/cpp/ArrowIterator). Will the CI test suite be enough?

cc @sfc-gh-sfan .

### CODE REVIEW

**Comment by 3dbrows** [comment]
Running `ci/build_docker.sh` produces a lot of C++ build errors so I imagine that does mean that the `ArrowIterator` would need some refactoring.

**Comment by noamcohen97** [comment]
created this branch that builds: #1349

**Comment by 3dbrows** [comment]
Fantastic, thank you @noamcohen97 , I see you upgraded to C++ 17 which is important as Arrow has updated to this recently. I will try including your C++ changes in this branch as well. The maintainers can then choose if they want to close this PR and accept only yours, or accept this one without the Python 3.11 change as well. I am happy either way.

Thank you again.

**Comment by noamcohen97** [comment]
I think if a new version is released, it might as well have new wheels built for 3.11

**Comment by noamcohen97** [comment]
> Fantastic, thank you @noamcohen97 , I see you upgraded to C++ 17 which is important as Arrow has updated to this recently. I will try including your C++ changes in this branch as well. The maintainers can then choose if they want to close this PR and accept only yours, or accept this one without the Python 3.11 change as well. I am happy either way.
> 
> Thank you again.

would really appreciate a co-authored-by on this :)
I am trying to build up my open source reputation

**Comment by 3dbrows** [comment]
btw @noamcohen97 I found I had to add `libarrow_dataset.so.1000` (and equivalents for other platforms) in order for `ci/build_docker.sh` to pass.

And yep, might as well do 3.11 at the same time.

> would really appreciate a co-authored-by on this :)
Yes, absolutely, how do I do this, edit a commit message I guess?

**Comment by noamcohen97** [comment]
> how do I do this, edit a commit message I guess
[Yes](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors)

> I found I had to add `libarrow_dataset.so.1000`

Nice!

**Comment by 3dbrows** [comment]
@noamcohen97 I think the attribution is fixed now :)

**Comment by joshuataylor** [comment]
Should it be >= 10.0.1 (for the Python 3.11 support)?

**Comment by 3dbrows** [comment]
Very well spotted @joshuataylor , thank you, I've bumped this to 10.0.1.

**Comment by nattaylor** [comment]
Perhaps this is a useless comment, but I checked out `3dbrows/pyarrow-10`, built on my M1, and I'm off and running.

File: setup.cfg [hunk 1]
```diff
@@ -91,6 +91,6 @@ development =
     pytzdata
 pandas =
     pandas>=1.0.0,<1.6.0
-    pyarrow>=8.0.0,<8.1.0
+    pyarrow>=10.0.1,<10.1.0
```
Comment (cpcloud) [inline setup.cfg:94]: Is there any particular reason multiple versions of arrow can't be supported?

Comment (joshuataylor) [reply setup.cfg:94]: With Python 3.11 only 10.0.1 and above is officially supported with prebuilt wheels.

Comment (sfc-gh-mkeller) [reply setup.cfg:94]: Major `pyarrow` version can change the ABI they use, which makes then not compatible with each other. Unfortunately; as long as we compile against `pyarrow` we can only support 1 major versions at a time.

**Comment by cpcloud** [comment]
Anything blocking merging this PR?

**Comment by 3dbrows** [comment]
@cpcloud Looks like it's just the description release notes that require updating - should these be updated in this PR, or is that done when the release is cut? If they should be updated now - under what (future) version number?

**Comment by sfc-gh-mkeller** [comment]
Merged through #1347

**Pull request closed without merging**